### ISSUE TITLE
V2 : Transforms are not applied in order

### ISSUE THREAD

**Issue by z3dev** [jscad/OpenJSCAD.org#520]
## Expected Behavior

Transforms (scale, rotate, translate, etc) are applied in order, resulting in expected shapes. The following code should produce three shapes, like this.

![good](https://user-images.githubusercontent.com/8969165/74507414-af409280-4f3f-11ea-8422-7a0e6cb33a60.png)

```
const { color, geometry, primitives, transforms } = require('@jscad/modeling')

const main = (params) => {
  let p0 = primitives.rectangle({center: [60.0000, -20.0000], size: [119.8, 39.8]})
  color.color([1, 0, 0, 1], p0)

  let p1 = primitives.roundedRectangle({center: [30.0000, -20.0000], size: [40, 20], roundRadius: 5})
  color.color([0, 1, 0, 1], p0)

  let p2 = primitives.roundedRectangle({center: [20.0000, -10.0000], size: [40, 20], roundRadius: 5})
  p2 = transforms.rotateZ(0.5235987755982988, p2)
  p2 = transforms.translate([70, -21, 0], p2)
  color.color([0, 0, 1, 1], p0)

  return [p0, p1, p2]
}

module.exports = { main }
```
## Actual Behavior

The transforms are accumulated, rather then being applied to the shape. This results in an unexpected shape.

![bad](https://user-images.githubusercontent.com/8969165/74507514-f4fd5b00-4f3f-11ea-8187-a72ddabd9ce2.png)

## Steps to Reproduce the Problem

  1. Drap and drop the given code into the web version of V2

## Specifications

  - Version: V2
  - Platform: ALL
  - Environment: ALL

**Comment by z3dev** [comment]
This is a BIG step backwards. Basically, users cannot predict the outcome of transforms.

Solution #1: Removed the lazy transforms in geom2/geom3/path2
Solution #2: Always apply transforms (is there any advantage to keeping the matrix?)

**Comment by z3dev** [comment]
Additional feedback... after changing the SVG deserializer to apply transforms as per SVG standards, the shapes are fine.

**Comment by kaosat-dev** [comment]
@z3dev I think this is our mistake in the order in which the transforms are applied, I need to double check a few things, but I might have a fix

**Comment by z3dev** [comment]
@kaosat-dev this may be just a learning issue, but there will be some fairly large differences between V1 and V2.

**Comment by kaosat-dev** [comment]
@z3dev I do not think it is a learning issue to be honest, you are completely right to point out that this is not the 'expected' result :)

**Comment by kaosat-dev** [comment]
hmm wait no , now I am not 100% sure hmm

**Comment by z3dev** [comment]
it's just different from V1. it's not wrong but different.

i'm curious. i'm sure transforms are part of the new V2 examplese. did you have any issues determining positions / rotations?

**Comment by z3dev** [comment]
Here's a nice tutorial about SVG transformations. I think V2 is following this pretty much.

**Comment by MatzeS** [comment]
Just a short notice:

`packages/modeling/src/geometry/geom2/transform.js` @ line 18, switching the order of operations does most of the trick.
`newgeometry.transforms = mat4.multiply(matrix, geometry.transforms)`

However, I am just trying to read up into the project so I have no idea what other implications this might have.
So, enjoy this suggestion with care and healty doubt.

**Comment by hjfreyer** [comment]
Agreed that this is backwards; I was about to file a bug myself.

The reason it feels backwards is because the order of matrix multiplications is the reverse of the function applications. If functions `f`, `g`, and `h` are transforms that apply matrices `F`, `G`, and `H`, then the syntax `f(g(h(geom)))` corresponds in linear-algebra-land to `H*G*F*geom` in the current implementation. People expect with the given syntax that `h` will be applied "first", then `g`, then `f`, because that's the convention in JS and most popular programming languages.

**Comment by hjfreyer** [comment]
Regarding the SVG documentation, the syntax there is different enough from JS that it may make sense in that context, but not in the JSCAD context. If JS's function call syntax looked like `geom |> f |> g |> h`, then I think the comparison to the SVG syntax would be more apt. But for better or worse, in JS, execution goes right to left, not left to right as in SVG.

**Comment by hjfreyer** [comment]
In fact, thinking about it further, this is _definitely_ wrong because it violates encapsulation.

Let's say we have a function `widget()` that returns a `geom3` with some transforms applied. Let's say it's a rotation of 90 degrees about the X axis (matrix `R`) followed by a translation (`T`), so the resulting geometry is `T*R*geom`. 

As a caller of this function, I should be able to take the output of `widget()` and transform it in ways that don't depend on its implementation (i.e., encapsulation). However, if I now try to translate it 10 units up along the Z axis using a matrix U, the resulting geometry will be  `T*R*U*geom`. 

The translation in the caller happens _before_ the rotation defined in `widget()`, so rather than moving the shape 10 units up, it'll move it 10 units along the Y axis. Likewise, if `widget()` had a matrix that scaled everything by 5x, then it'd move 50 units. You can't anticipate that unless you look at the code of `widget()`, and `widget()` can't be safely changed to return a different representation of the same geometry, because that could break callers. In other words, functions don't encapsulate their implementations.

Definitely should be changed.

**Comment by z3dev** [comment]
This should be fixed with the merge of #644 so closing this issue.
